## IMPORTANT POINTS

## - In this task i was asked to create a tool to predict the disease of Ms Sara & i was asked to tell the likelihood of it being Covid or Malaria but I ADDED MORE DISEASE

## - it is mentioned in task that she had only two symptoms but this program can intake 7 SYMPTOMS as inputs

## - only covid symptom dataset was available online & malaria's dataset contained microscopic images only

## - i created malaria's dataset mannually by studying it's symptoms online which was quite time consuming

## IMPORTING RELEVANT LIBRARIES

In [149]:
import numpy as np
import pandas as pd

## LOADING THE DATA

In [150]:
df1 = pd.read_excel('dataset.xlsx')

## CHECKING THE DATAHEAD

In [151]:
df1.head()

cough  fever  sore_throat  shortness_of_breath  head_ache  vomiting  \
0      1      0            0                    0          0         0   
1      0      0            0                    0          0         0   
2      1      0            0                    0          0         0   
3      0      1            0                    0          1         0   
4      0      1            0                    0          0         0   

   chills  covid  malaria  flu  
0       0      1        0    0  
1       0      1        0    0  
2       0      1        0    0  
3       0      1        0    0  
4       0      1        0    0

## SHUFFLING THE DATA

In [152]:
df2=df1.sample(frac=1)

In [153]:
df2.head()

cough  fever  sore_throat  shortness_of_breath  head_ache  vomiting  \
669       0      0            0                    0          0         0   
2516      0      0            0                    0          0         0   
2615      0      0            0                    0          0         0   
2523      0      0            0                    0          0         0   
17        1      0            0                    0          0         0   

      chills  covid  malaria  flu  
669        0      1        0    0  
2516       1      0        0    1  
2615       0      0        0    1  
2523       1      0        0    1  
17         0      1        0    0

## IMPORTING SKLEARN TO SPLIT THE DATASET

In [154]:
from sklearn.model_selection import train_test_split
from sklearn.metrics import accuracy_score

## ASSIGNING THE INPUT AND TARGET COLUMNS

In [155]:
x = df2.drop(columns = ['covid','malaria','flu'],axis=1)
k = df2[['covid','malaria','flu']]

In [156]:
x

cough  fever  sore_throat  shortness_of_breath  head_ache  vomiting  \
669       0      0            0                    0          0         0   
2516      0      0            0                    0          0         0   
2615      0      0            0                    0          0         0   
2523      0      0            0                    0          0         0   
17        1      0            0                    0          0         0   
...     ...    ...          ...                  ...        ...       ...   
507       0      0            0                    0          0         0   
1209      0      0            0                    0          1         1   
385       0      0            0                    0          0         0   
2407      0      0            0                    0          0         0   
1559      0      0            0                    0          0         0   

      chills  
669        0  
2516       1  
2615       0  
2523       1  
17         0  
...      ...  
507        0  
1209       0  
385        0  
2407       1  
1559       0  

[3000 rows x 7 columns]

In [157]:
k

covid  malaria  flu
669       1        0    0
2516      0        0    1
2615      0        0    1
2523      0        0    1
17        1        0    0
...     ...      ...  ...
507       1        0    0
1209      0        1    0
385       1        0    0
2407      0        0    1
1559      0        1    0

[3000 rows x 3 columns]

## IMPORTING RELEVANT LIBRARIES

In [158]:
from tensorflow.keras.models import Sequential
from tensorflow.keras.layers import Dense
from tensorflow.keras.losses import binary_crossentropy
from tensorflow.keras.optimizers import Adam

## SELECTING FEATURES

In [168]:
n_samples = 3000
n_features = 7
n_classes = 3
n_labels = 2
n_epochs = 50
random_state = 42
batch_size = 250
verbosity = 1
validation_split = 0.2

## TRAINING AND TESTING

In [169]:
X_train, X_test, y_train, y_test = train_test_split(x, k, test_size=0.33,stratify = k,random_state=random_state)

## CREATING MODEL

In [170]:
model = Sequential()
model.add(Dense(32, activation='relu', input_dim=n_features))
model.add(Dense(16, activation='relu'))
model.add(Dense(8, activation='relu'))
model.add(Dense(n_classes, activation='sigmoid'))

## COMPILING THE MODEL

In [171]:
model.compile(loss=binary_crossentropy,
              optimizer=Adam(),
              metrics=['accuracy'])

## FITTING THE TRAIN & TEST DATA

In [172]:
model.fit(X_train, y_train,
          batch_size=batch_size,
          epochs=n_epochs,
          verbose=verbosity,
          validation_split=validation_split)

Epoch 1/50
7/7 [==============================] - 5s 140ms/step - loss: 0.6885 - accuracy: 0.2326 - val_loss: 0.6826 - val_accuracy: 0.4080
Epoch 2/50
7/7 [==============================] - 0s 42ms/step - loss: 0.6802 - accuracy: 0.3924 - val_loss: 0.6742 - val_accuracy: 0.4428
Epoch 3/50
7/7 [==============================] - 0s 62ms/step - loss: 0.6718 - accuracy: 0.4136 - val_loss: 0.6647 - val_accuracy: 0.4403
Epoch 4/50
7/7 [==============================] - 0s 71ms/step - loss: 0.6627 - accuracy: 0.4266 - val_loss: 0.6544 - val_accuracy: 0.5075
Epoch 5/50
7/7 [==============================] - 0s 55ms/step - loss: 0.6527 - accuracy: 0.3644 - val_loss: 0.6438 - val_accuracy: 0.3433
Epoch 6/50
7/7 [==============================] - 0s 67ms/step - loss: 0.6421 - accuracy: 0.3327 - val_loss: 0.6338 - val_accuracy: 0.3433
Epoch 7/50
7/7 [==============================] - 0s 38ms/step - loss: 0.6327 - accuracy: 0.3327 - val_loss: 0.6241 - val_accuracy: 0.3433
Epoch 8/50
7/7 [==========

## CHECKING FOR ACCURACY

In [173]:
score = model.evaluate(X_test, y_test, verbose=0)

In [174]:
print(f'Test loss: {score[0]} / Test accuracy: {score[1]}')

Test loss: 0.3836374580860138 / Test accuracy: 0.6898989677429199


## PREDICTING DISEASE ON THE BASIS OF GIVEN SYMPTOMS (FEVER & HEADACHE)

In [175]:
cf = [] # CREATING A LIST TO STORE INPUT VALUES
y = 0 
while len(cf)<7:
    z = x.columns[y] # TO PRINT THE NAME OF THE SYMPTOM IN PROMPT WHICH IS ALSO THE NAME OF THE COLUMN
    print('Do You Have:',z) # ASKING FOR SYMPTOMS
    symp = input('Enter 1 = Yes,0 = No:') # 1 MEANS YES AND 0 MEANS NO
    f = symp.split(",") # CONVERTING IT INTO LIST
    cf.append(f[0]) # APPENDING IT INTO LIST
    y += 1
    for i in range(0, len(cf)): # CONVERTING STRING LIST INTO INTEGER LIST
        cf[i] = int(cf[i])
        
tr = model.predict([cf]).round(2) # MAKING PREDICTIONS AND ROUNDING OFF THE FIGURE UPTO 2 DECIMAL PLACES
    
hlist = [] # CREATING A LIST TO STORE THE PREDICTED VALUES
for i in tr: # LOOPING TO EXTRACT VALUES
    i = i*100 # MULTIPLYING BY 100 TO MAKE THIS PERCENTAGE IN HUNDREDS NOT IN DECIMALS
    hlist.append(i[0])
    hlist.append(i[1])
    hlist.append(i[2])
    df_fn = pd.DataFrame()
    df_fn['Disease'] = ['Covid','Malaria','Flu'] # CREATING A NEW DATAFRAME AND NAMING THE COLUMNS WITH THE REQUIRED DISEASE
    df_fn['Likelihood_In_Percent'] = hlist # STORING THE DATA IN THE APPROPRIATE ROW OF THIS NEWLY FORMED DATAFRAME
    final = df_fn.sort_values(by='Likelihood_In_Percent', ascending=False) # PRINTING THE DATA IN DESCENDING ORDER AS REQUIRED
    print(final)

Do You Have: cough
Enter 1 = Yes,0 = No:0
Do You Have: fever
Enter 1 = Yes,0 = No:1
Do You Have: sore_throat
Enter 1 = Yes,0 = No:0
Do You Have: shortness_of_breath
Enter 1 = Yes,0 = No:0
Do You Have: head_ache
Enter 1 = Yes,0 = No:1
Do You Have: vomiting
Enter 1 = Yes,0 = No:0
Do You Have: chills
Enter 1 = Yes,0 = No:0
   Disease  Likelihood_In_Percent
1  Malaria                   56.0
0    Covid                   20.0
2      Flu                    2.0


## CONCLUSION:- On the basis of given symptoms we found out that there is 56% chance that Ms Sara has MALARIA, 20% chance for COVID & 2% chance for FLU.......BUT WE CAN PUT MORE SYMPTOMS IN THAT PROMPT FOR MORE ACCURACY